### PR TITLE
Drop miniupnp dependency

### PR DESCRIPTION
**Pull request by darosior**
This PR removes UPnP IGD support and drops our [miniupnp](https://github.com/miniupnp/miniupnp) dependency.

Miniupnpc is a C library (somewhat) maintained by a single person which had several vulnerabilities in the past (a couple dozens are listed [here](https://cve.mitre.org/cgi-bin/cvekey.cgi?keyword=miniupnp)), some of which directly affected our software ([RCE in 2015](https://bitcoincore.org/en/2024/07/03/disclose_upnp_rce/), [OOM in 2020](https://bitcoincore.org/en/2024/07/31/disclose-upnp-oom/)).

The main purpose of this functionality is to have more (non-data-center) reachable nodes on the network. For a non-technical user running Bitcoin Core at home, the software would automatically open a port on their router to receive incoming connections. This way, users not able to manually open a port on their router would still provide the network with more resources and enhance its diversity.

However, due to past vulnerabilities (and a worry about unknown future ones) in miniupnpc this feature was disabled by default in https://github.com/bitcoin/bitcoin/pull/6795. Having it disabled by default kills (most of?) the purpose of having this functionality in the first place: someone technical enough to understand the `-upnp` startup option or the "enable UPnP" setting is most likely able to open a port on his box in the first place.

In addition, laanwj implemented PCP with a NAT-PMP fallback directly in Bitcoin Core in https://github.com/bitcoin/bitcoin/pull/30043. If we ever want to re-enable automatic NAT traversal by default in Bitcoin Core, this is the best option (and in my opinion the only sane one). The NAT-PMP fallback makes it so compatibility shouldn't be (much of) an issue.

On balance, i believe that keeping this functionality and this barely maintained C dependency has higher costs than benefits. Therefore i propose that we get rid of it.

### CODE REVIEW

**Review by stickies-v — COMMENTED** [review]
Concept ACK

**Comment by kevkevinpal** [comment]
Concept ACK 

There have been instances where there have been breakage when using newer minor versions of miniupnp aswell to the other mentioned reasons
https://github.com/bitcoin/bitcoin/issues/30266 -> https://github.com/bitcoin/bitcoin/pull/30283 -> https://github.com/bitcoin/bitcoin/pull/30301

**Comment by kevkevinpal** [comment]
Might be worth adding a release note?

**Comment by fjahr** [comment]
Concept ACK but shouldn't this go through a deprecation cycle?

File: src/init.cpp [hunk 1]
```diff
@@ -561,11 +561,6 @@ void SetupServerArgs(ArgsManager& argsman, bool can_listen_ipc)
...
     argsman.AddArg("-torcontrol=<ip>:<port>", strprintf("Tor control host and port to use if onion listening enabled (default: %s). If no port is specified, the default port of %i will be used.", DEFAULT_TOR_CONTROL, DEFAULT_TOR_CONTROL_PORT), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
     argsman.AddArg("-torpassword=<pass>", "Tor control port password (default: empty)", ArgsManager::ALLOW_ANY | ArgsManager::SENSITIVE, OptionsCategory::CONNECTION);
-#ifdef USE_UPNP
-    argsman.AddArg("-upnp", strprintf("Use UPnP to map the listening port (default: %u)", DEFAULT_UPNP), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
-#else
-    hidden_args.emplace_back("-upnp");
-#endif
     argsman.AddArg("-natpmp", strprintf("Use PCP or NAT-PMP to map the listening port (default: %u)", DEFAULT_NATPMP), ArgsManager::ALLOW_ANY, OptionsCategory::CONNECTION);
```
Comment (laanwj) [inline src/init.cpp:564]: Might want to keep this arg around for one major version, but make it display an error that the functionality has been removed. This gives a clearer error to anyone who has it in their configuration file than "argument doesn't exist".

**Comment by laanwj** [comment]
> Concept ACK but shouldn't this go through a deprecation cycle?

imo a real deprecation cycle is not needed here. This has teetered on the brink of being removed for so long already. Also, deprecation cycles are generally for APIs, and this doesn't add or remove any RPC or REST functionality.

**Comment by Christewart** [comment]
concept ACK

**Comment by edilmedeiros** [comment]
Concept ACK

File: src/init.cpp (hunk 1 above)
Comment (darosior) [reply src/init.cpp:565]: Done.

**Comment by darosior** [comment]
Kept the option as a hidden arg which will return an error explaining UPnP support was dropped and encouraging users to set `-natpmp` instead.

Added a release note.

**Comment by 1440000bytes** [comment]
> Miniupnpc is a C library (somewhat) maintained by a single person which had several vulnerabilities in the past (a couple dozens are listed here), some of which directly affected our software (RCE in 2015, OOM in 2020).

and CVE-2017-1000494 in 2018?

Concept ACK 

Lint error: https://github.com/bitcoin/bitcoin/pull/31130/checks?check_run_id=31973806317

File: doc/release-notes-31130.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+P2P and network changes
+-----------------------
+
+Support for UPnP was dropped. Consider using PCP or NAT-PMP instead.
```
[the remark this answers is not in the document]
Comment (laanwj) [reply doc/release-notes-31130.md:4]: maybe
"If you want to open a port automatically, consider using the `-natpmp` option instead, which uses PCP or NAT-PMP depending on router support."
(in the GUI, the option is still called NATPMP as well, so no need to mention that seperately i think)

[the remark this answers is not in the document]
Comment (darosior) [reply doc/release-notes-31130.md:4]: Took the suggestion, thanks.

**Comment by darosior** [comment]
Updated the release note to something clearer and fixed the `check-doc.py` lint error by using `AddArg()` directly for the hidden `-upnp` arg.

**Comment by darosior** [comment]
Rebased after #31148 merge.

**Comment by pythcoiner** [comment]
concept ACK

**Comment by i-am-yuvi** [comment]
Concept ACK!!

**Review by maflcko — APPROVED** [review]
Almost lgtm. Didn't review the gui changes and the rebase was borked.

ACK 2a541e93cf13362e52c62e4d01d34a602843471e 📏

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 2a541e93cf13362e52c62e4d01d34a602843471e 📏
5Gd8ObbrtS4VtkpC0GlzanymPWZ6n1+iRsigBUfaI+d+RC5UmLKR6iPjc6MOit/UP1GhKd52edRMDiWGb3rsAg==
```

</details>

File: src/mapport.cpp [hunk 3]
```diff
@@ -268,7 +176,7 @@ static void DispatchMapPort()
 
     assert(g_mapport_thread.joinable());
     assert(!g_mapport_interrupt);
-    // Interrupt a protocol-specific loop in the ThreadUpnp() or in the ThreadPCP()
```
Comment (fanquake) [inline src/mapport.cpp:271]: I think you can just drop this comment entirely, as there is no-longer a "next protocol" in the loop.

File: src/mapport.cpp [hunk 4]
```diff
@@ -219,15 +136,6 @@ static void ThreadMapPort()
             if (ok) continue;
         }
 
-#ifdef USE_UPNP
```
Comment (fanquake) [inline src/mapport.cpp:222]: Just up from here, could drop `// High priority protocol.`.

File: src/mapport.cpp [hunk 5]
```diff
@@ -268,7 +176,7 @@ static void DispatchMapPort()
```
Comment (fanquake) [inline src/mapport.cpp:176]: Above here:
```cpp
    if (g_mapport_enabled_protos & g_mapport_current_proto) {
        // Enabling another protocol does not cause switching from the currently used one.
        return;
    }
```
This code is either dead, or the comment is outdated, as we can no-longer "enable another protocol".

**Review by fanquake — COMMENTED** [review]
Not sure how much you want to do here, but I think this code can be cleaned up further, given it was written to cater for using two protocols, but after this change, we've got 1 protocol (with a fallback).

File: src/mapport.cpp (hunk 3 above)
Comment (laanwj) [reply src/mapport.cpp:271]: i think cleaning up a bit more makes sense here, though let's keep it easy to review refactor-wise

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -67,12 +67,12 @@ jobs:
...
       - name: Compile and run tests
         run: |
           # Run tests on commits after the last merge commit and before the PR head commit
           # Use clang++, because it is a bit faster and uses less memory than g++
-          git rebase --exec "echo Running test-one-commit on \$( git log -1 ) && CC=clang CXX=clang++ cmake -B build -DWERROR=ON -DWITH_ZMQ=ON -DBUILD_GUI=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DWITH_BDB=ON -DWITH_MINIUPNPC=ON -DWITH_USDT=ON && cmake --build build -j $(nproc) && ctest --output-on-failure --test-dir build -j $(nproc) && ./build/test/functional/test_runner.py -j $(( $(nproc) * 2 ))" ${{ env.TEST_BASE }}
+          git rebase --exec "echo Running test-one-commit on \$( git log -1 ) && CC=clang CXX=clang++ cmake -B build -DWERROR=ON -DWITH_ZMQ=ON -DBUILD_GUI=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DWITH_BDB=ON -DWITH_USDT=ON && cmake --build build -j $(nproc) && ctest --test-dir build -j $(nproc) && ./build/test/functional/test_runner.py -j $(( $(nproc) * 2 ))" ${{ env.TEST_BASE }}
```
[the remark this answers is not in the document]
Comment (darosior) [reply .github/workflows/ci.yml]: My bad i did this one in a hurry. Fixed now, thanks for catching this.

**Comment by darosior** [comment]
I know this can be cleaned up, but i wanted to keep it minimal and focused on dropping UPnP. I'm happy to do a few cleanups (such as the ones you suggest) but i don't want to start refactoring the whole file. It can always happen in a follow-up if someone is interested and more familiar with this code.

**Comment by fanquake** [comment]
> I know this can be cleaned up, but i wanted to keep it minimal and focused on dropping UPnP. I'm happy to do a few cleanups (such as the ones you suggest) but i don't want to start refactoring the whole file. 

Sure, that's fine, any more-invasive changes could be done later, but we shouldn't be leaving around dead code, or commentary that is incorrect.

**Comment by laanwj** [comment]
>  i don't want to start refactoring the whole file. It can always happen in a follow-up if someone is interested and more familiar with this code.

i agree. Let's make sense there are no clearly wrong comments, or dead code. But as we're sure that the only port mapping method we want to keep is PCP/NATPMP, a lot of the logic in this file can be removed and methods can be renamed. But might as well do that in a future follow-up to #30043, as some of those need larger restructures to the loop too (to keep state between runs).

**Comment by theuni** [comment]
Getting on the Concept ACK train. Another dependency gone!

Agree with @laanwj about the refactor as a separate PR.

**Comment by edilmedeiros** [comment]
utACK b7b24352906f1dba64826e7a093069b5bfc504dc.

Agree with @fanquake comments.

File: src/mapport.cpp (hunk 3 above)
Comment (darosior) [reply src/mapport.cpp:271]: Done.

File: src/mapport.cpp (hunk 5 above)
Comment (darosior) [reply src/mapport.cpp]: It was both. Removed outdated comments and dead code in the latest push. I cleaned up this function in the follow-up.

**Comment by darosior** [comment]
Pushed two commits to remove outdated comments and dead code. I opened https://github.com/bitcoin/bitcoin/pull/31157 as a follow-up to clean up some of the logic which became unnecessary or confusing now that there is only a single protocol option.

**Review by 1440000bytes — APPROVED** [review]
Code Review ACK https://github.com/bitcoin/bitcoin/pull/31130/commits/40e5f26a3ff77e50df808f6f850c617aec2df203

**Review by laanwj — APPROVED** [review]
Code review ACK 40e5f26a3ff77e50df808f6f850c617aec2df203
i have also checked that:
- no mentions of miniupnp as a dependency remain
- `-natpmp` still works to forward ports on my router after this
- `-upnp`, if still present in the configuration file or command line, gives a custom message and exits

**Review by yuvicc — APPROVED** [review]
Tested ACK 40e5f26a3ff77e50df808f6f850c617aec2df203

- -upnp gives custom messages and exits
- -natpmp works

**Comment by fanquake** [comment]
Guix Build:
```bash
541e5a857759e6e34aa8020e70b2b89be7cadfb358aecb0d83c10df214c5b1f7  guix-build-40e5f26a3ff7/output/aarch64-linux-gnu/SHA256SUMS.part
3973c1c93d969f4b1a338c734df151db3bdfdca0fcb5800f126890a426d5b35d  guix-build-40e5f26a3ff7/output/aarch64-linux-gnu/bitcoin-40e5f26a3ff7-aarch64-linux-gnu-debug.tar.gz
4cdeb68207b4f71819618b53f1c463f368ccca2912617945caceb49049f30a80  guix-build-40e5f26a3ff7/output/aarch64-linux-gnu/bitcoin-40e5f26a3ff7-aarch64-linux-gnu.tar.gz
5da6d6e1ab2ac471dfcaf9276f54d3a81ed951700d78354d9b3431f7954e70f4  guix-build-40e5f26a3ff7/output/arm-linux-gnueabihf/SHA256SUMS.part
4d8126f08b2474408dbc72c44c557397edaa318b6abc43bbb27fbc0ff9081420  guix-build-40e5f26a3ff7/output/arm-linux-gnueabihf/bitcoin-40e5f26a3ff7-arm-linux-gnueabihf-debug.tar.gz
dc04c665dd446426532335a9049e93f739031db7568e6f17a7cb9f45c99d9687  guix-build-40e5f26a3ff7/output/arm-linux-gnueabihf/bitcoin-40e5f26a3ff7-arm-linux-gnueabihf.tar.gz
4ea112d12ac5c0c527fcef92bf02cba4f56ff75296131eb20b4b28c134460cd4  guix-build-40e5f26a3ff7/output/arm64-apple-darwin/SHA256SUMS.part
39ebe16080b018d1ddb20c486f66a5e6180da02670a8fcc91dcf2464bbe29e9d  guix-build-40e5f26a3ff7/output/arm64-apple-darwin/bitcoin-40e5f26a3ff7-arm64-apple-darwin-unsigned.tar.gz
52ed74eb097078de8a8b68233d113ad826f9e937d71081733e4ecd5dcbdf1024  guix-build-40e5f26a3ff7/output/arm64-apple-darwin/bitcoin-40e5f26a3ff7-arm64-apple-darwin-unsigned.zip
b5a67473c600ccca0f5416939b9e1eebc95caa624522fede3319a31f0276bba7  guix-build-40e5f26a3ff7/output/arm64-apple-darwin/bitcoin-40e5f26a3ff7-arm64-apple-darwin.tar.gz
9049e160d1cbb7ba3f616ee4ef4078ebbb7580af1feb03f1eedddaf17e3f574c  guix-build-40e5f26a3ff7/output/dist-archive/bitcoin-40e5f26a3ff7.tar.gz
d1bd12e1450bcd1e513a9b5bf6721d9a91da54db116651eae135f36cfd3bc3c5  guix-build-40e5f26a3ff7/output/powerpc64-linux-gnu/SHA256SUMS.part
38dacbe2edd48f9b3e54835cb4c15ebc0288abc80ae2e4d2f0376df2c9c64c33  guix-build-40e5f26a3ff7/output/powerpc64-linux-gnu/bitcoin-40e5f26a3ff7-powerpc64-linux-gnu-debug.tar.gz
4170b6b3ac4ee6cd2aa26c7767831469fd83dc2d1fe18145ec425cbcf917a38f  guix-build-40e5f26a3ff7/output/powerpc64-linux-gnu/bitcoin-40e5f26a3ff7-powerpc64-linux-gnu.tar.gz
6aeebc67d7052641176bfb7cd05a3c69b50780e97aebd2793db8bdd0c0a524d3  guix-build-40e5f26a3ff7/output/riscv64-linux-gnu/SHA256SUMS.part
393e5c2429d78c3fc9183a6cb8b9c56c98b01d45a8a85bd14d2ff1f73088626d  guix-build-40e5f26a3ff7/output/riscv64-linux-gnu/bitcoin-40e5f26a3ff7-riscv64-linux-gnu-debug.tar.gz
0103ad4699282b908cbba54afd82828ca9b30a8e13f447a4a6c2c50275835203  guix-build-40e5f26a3ff7/output/riscv64-linux-gnu/bitcoin-40e5f26a3ff7-riscv64-linux-gnu.tar.gz
c4c66663a8280577e0f116eaa42522d93a8535144018d9eb40f7a8b96bff636c  guix-build-40e5f26a3ff7/output/x86_64-apple-darwin/SHA256SUMS.part
75e0e43cdb0842c57f4b58ba100f8f5270137c236ac9c83c0586a8041df90729  guix-build-40e5f26a3ff7/output/x86_64-apple-darwin/bitcoin-40e5f26a3ff7-x86_64-apple-darwin-unsigned.tar.gz
8f23f92e1e542c5c5c933ef6e8fad89fe0a667946d8630dbbc9090425aa4b605  guix-build-40e5f26a3ff7/output/x86_64-apple-darwin/bitcoin-40e5f26a3ff7-x86_64-apple-darwin-unsigned.zip
f8b109eb63bd1cb37a18f664a123c3dce206394f3649510330efab33e0e2bdfd  guix-build-40e5f26a3ff7/output/x86_64-apple-darwin/bitcoin-40e5f26a3ff7-x86_64-apple-darwin.tar.gz
9b38c99dedbc0f7841587532cb7ccaf899006f9fcfe0f7bd35999aa6775bc41d  guix-build-40e5f26a3ff7/output/x86_64-linux-gnu/SHA256SUMS.part
65efd9a3fd967069db09c2e705449b8223809ac1d4e1f62f312eaeb059122ee6  guix-build-40e5f26a3ff7/output/x86_64-linux-gnu/bitcoin-40e5f26a3ff7-x86_64-linux-gnu-debug.tar.gz
4a9439f81cb98dc3943d8bc3c14dd3fa96dafe20401b8c8be97ee97ee0bd996c  guix-build-40e5f26a3ff7/output/x86_64-linux-gnu/bitcoin-40e5f26a3ff7-x86_64-linux-gnu.tar.gz
52a47e5908b85a1324e5c2d90dd55c188cb0712495a32801bd201adab6b4490f  guix-build-40e5f26a3ff7/output/x86_64-w64-mingw32/SHA256SUMS.part
186521473489789e6e2cd3bb4f9c3e901168282305b99c9e66741caffa8db4d8  guix-build-40e5f26a3ff7/output/x86_64-w64-mingw32/bitcoin-40e5f26a3ff7-win64-debug.zip
ceb9204a1c643b1aee480ed5fb24da436713753fa1ff966124174908743db342  guix-build-40e5f26a3ff7/output/x86_64-w64-mingw32/bitcoin-40e5f26a3ff7-win64-setup-unsigned.exe
f816302f2023d42df33fd3896f380c4734d2333ff7e8bb06004926aabeb25128  guix-build-40e5f26a3ff7/output/x86_64-w64-mingw32/bitcoin-40e5f26a3ff7-win64-unsigned.tar.gz
440d575cc18508ad40e9959ca5e479216232a76ecd46dc96d3a7528dee145475  guix-build-40e5f26a3ff7/output/x86_64-w64-mingw32/bitcoin-40e5f26a3ff7-win64.zip
```

**Comment by Sjors** [comment]
The GUI needs to handle this more gracefully:

Scherm­afbeelding 2024-10-30 om 12 13 44

A user would have to figure out that they need to manually edit `settings.json` or delete it and redo all their settings.

We should probably automatically delete it from `settings.json`. And then display a warning / info (instead of error) that tells users to consider the PCP setting in the GUI.

**Comment by laanwj** [comment]
> We should probably automatically delete it from settings.json. And then display a warning / info (instead of error) that tells users to consider the PCP setting in the GUI.

> I posted about this on irc but no one commented on it :(

Yes. Please create a new issue for this

**Comment by Sjors** [comment]
Done in https://github.com/bitcoin-core/gui/issues/843

File: src/qt/optionsmodel.cpp [hunk 7]
```diff
@@ -789,7 +776,6 @@ void OptionsModel::checkAndMigrate()
     migrate_setting(SpendZeroConfChange, "bSpendZeroConfChange");
     migrate_setting(ExternalSignerPath, "external_signer_path");
 #endif
-    migrate_setting(MapPortUPnP, "fUseUPnP");
```
Comment (ryanofsky) [inline src/qt/optionsmodel.cpp:792]: In commit "qt: remove UPnP settings" (844770b05ebc34789dc46d70cd6398089539c915)

It probably would be better to replace this line with:

```c++
if (settings.contains("fUseUPnP")) {
    if (node().getPersistentSetting("upnp").isNull()) UpdateRwSetting(node(), "upnp", "", settings.value("fUseUPnP").toBool());
    settings.remove("fUseUPnP");
}
```

So the setting would still be migrated and a warning could be logged if it was enabled.

**Pull request merged**